### PR TITLE
[SPARK-8479] [MLlib] Add numNonzeros and numActives to linalg.Matrices

### PR DESCRIPTION
**Pull request by MechCoder**
Matrices allow zeros to be stored in values. Sometimes a method is handy to check if the numNonZeros are same as number of Active values.

### CODE REVIEW

File: project/MimaExcludes.scala [hunk 1]
```diff
@@ -176,6 +176,12 @@ object MimaExcludes {
             ProblemFilters.exclude[MissingClassProblem]("org.apache.spark.sql.jdbc.PostgresQuirks"),
             ProblemFilters.exclude[MissingClassProblem]("org.apache.spark.sql.jdbc.NoQuirks"),
             ProblemFilters.exclude[MissingClassProblem]("org.apache.spark.sql.jdbc.MySQLQuirks")
+          ) ++ Seq(
+            // SPARK-8479 Add numNonzeros and numActives to Matrix.
+            ProblemFilters.exclude[MissingClassProblem](
```
Comment (mengxr) [inline project/MimaExcludes.scala]: Should it be "MissingMethodProblem" instead?

**Comment by MechCoder** [comment]
@mengxr Sorry about that. Btw, when is a foreach preferred as compared to a while loop.

**Comment by MechCoder** [comment]
ping @srowen Could you please help me figure out the compatibility problems here? I've added it to the filters but it does not seem to work.

File: mllib/src/main/scala/org/apache/spark/mllib/linalg/Matrices.scala [hunk 2]
```diff
@@ -592,6 +614,19 @@ class SparseMatrix(
   def toDense: DenseMatrix = {
     new DenseMatrix(numRows, numCols, toArray)
   }
+
+  override def numNonzeros: Int = {
+    var nnz = 0
```
Comment (srowen) [inline mllib/src/main/scala/org/apache/spark/mllib/linalg/Matrices.scala]: Can this not be `values.count(_ != 0)`?

**Comment by srowen** [comment]
I think the failure is indeed because you're adding methods to a trait, but it's a sealed trait, so probably fine to override. I think you put it in the wrong place? it shoudl go under 1.5

**Comment by MechCoder** [comment]
Thanks for the comment but Vector is a sealed trait as well but that seems to work fine.

**Comment by mengxr** [comment]
MiMa doesn't know it. So we still need to add the exclusion rules explicitly, otherwise it won't pass Jenkins.

**Comment by MechCoder** [comment]
Have I done it correctly?

**Comment by mengxr** [comment]
LGTM. Merged into master. Thanks!

**Pull request closed without merging**